### PR TITLE
Lazy create MVC ControllerActionInvoker statemachines

### PR DESCRIPTION
**Pull request by benaadams**
* Lazy create ControllerActionInvoker.InvokeInnerFilterAsync statemachine
* Lazy create ControllerActionInvoker.InvokeNextActionFilterAwaitedAsync statemachine
* Lazy create ControllerActionInvoker.InvokeNextActionFilterAsync statemachine
* Lazy create ControllerActionInvoker.InvokeActionMethodAsync statemachine
* Lazy create Controller.OnActionExecutionAsync statemachine

image

/cc @rynowak  @pranavkm

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Infrastructure/ControllerActionInvoker.cs [hunk 1]
```diff
@@ -320,69 +369,119 @@ private async Task<ActionExecutedContext> InvokeNextActionFilterAwaitedAsync()
...
+            else
+            {
+                return Awaited(this, actionResultValueTask);
+            }
 
-            IActionResult result = null;
-            try
+            return Task.CompletedTask;
+
+            static async Task Awaited(ControllerActionInvoker invoker, ValueTask< IActionResult> actionResultValueTask)
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/ControllerActionInvoker.cs]: Fix da spacing :)

Comment (benaadams) [reply src/Mvc/Mvc.Core/src/Infrastructure/ControllerActionInvoker.cs]: Fixed :)

**Comment by benaadams** [comment]
Update on outstanding statemachines with current sdk; highlighting Controller stuff (this PR)
image

**Review by rynowak — APPROVED** [review]

**Comment by rynowak** [comment]
Thanks!

**Comment by rynowak** [comment]
When all of this stuff is done it would be super cool to see before/after numbers

**Comment by benaadams** [comment]
> When all of this stuff is done it would be super cool to see before/after numbers

3.0 is currently +37.3% higher than 2.2

image


image

**Comment by benaadams** [comment]
@rynowak has stabilised for now as nothing merged related, though still https://github.com/aspnet/AspNetCore/pull/9512 and https://github.com/aspnet/Extensions/pull/1479 to go

image

Drop was from GC issue that @sebastienros tracked down and has now bee resolved.

**Pull request merged**